openedx/openedx-platform
# [BD-04] Convert Conditional XModule to XBlock

Converts the Conditional XModule into an XBlock.

Part of [XModule to XBlock Conversion work](https://openedx.atlassian.net/wiki/spaces/AC/pages/1472790755/XModule+to+XBlock+Conversion).

**Testing Instructions**
1. Create a course. Add `conditional` to the Advanced Module List and then create a few conditional modules. See the [docs](https://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/latest/exercises_tools/conditional_module.html#conditional-module) for further details.
2. Export the course.
3. Update devstack with this code.
   - Cherry-pick this commit.
   - In both `make lms-shell` and `make studio-shell`, install the updated version of the `xmodule` package with `pip install -e /common/lib/xmodule`.
   - Do `make lms-static` and `make studio-static`.
   - Do `make lms-restart` and `make studio-restart`.
4. Test the existing components in both Studio and LMS to verify the behavior is the same.
5. Import the exported archive into a new course and check that components are imported and work correctly.

**MRO Analysis**
Descriptor MRO | Module MRO | Block MRO | Notes
------------ | ------------- | ---------------- | ------------
<class 'xmodule.conditional_module.ConditionalDescriptor'> | <class 'xmodule.conditional_module.ConditionalModule'> | <class 'xmodule.conditional_module.ConditionalBlock'> | Merged into one class.
<class 'xmodule.conditional_module.ConditionalFields'> | <class 'xmodule.conditional_module.ConditionalFields'> | — | Merged into ConditionalBlock.
<class 'xmodule.seq_module.SequenceDescriptor'> | — | — | The attributes needed by ConditionalBlock have been seperated into SequenceMixin
<class 'xmodule.seq_module.SequenceMixin'> | — | <class 'xmodule.seq_module.SequenceMixin'> | —
<class 'xmodule.seq_module.SequenceFields'> | — | <class 'xmodule.seq_module.SequenceFields'> | —
<class 'xmodule.seq_module.ProctoringFields'> | — | — | Not needed by ConditionalBlock.
<class 'xmodule.mako_module.MakoModuleDescriptor'> | — | — | Inherits from (MakoTemplateBlockBase, XModuleDescriptor) and has the get_html() method which has been implemented in the XBlock.
<class 'xmodule.mako_module.MakoTemplateBlockBase'> | — | <class 'xmodule.mako_module.MakoTemplateBlockBase'> | —
<class 'xmodule.xml_module.XmlDescriptor'> | — | — | Inherits from (XmlMixin, XModuleDescriptor) and doens't have any atttributes so we can remove it.
<class 'xmodule.xml_module.XmlMixin'> | — | <class 'xmodule.xml_module.XmlMixin'> | —
<class 'xmodule.xml_module.XmlParserMixin'> | — | <class 'xmodule.xml_module.XmlParserMixin'> | —
<class 'xmodule.x_module.XModuleDescriptor'> | — | — | Not needed.
<class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | — | <class 'xmodule.x_module.XModuleDescriptorToXBlockMixin'> | —
— | <class 'xmodule.x_module.XModule'> | — | Not needed.
— | <class 'xmodule.x_module.XModuleToXBlockMixin'> | <class 'xmodule.x_module.XModuleToXBlockMixin'> | —
<class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | <class 'xmodule.x_module.HTMLSnippet'> | —
<class 'xmodule.x_module.ResourceTemplates'> | — | <class 'xmodule.x_module.ResourceTemplates'> | —
<class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | <class 'xmodule.x_module.XModuleMixin'> | —
<class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | <class 'xmodule.x_module.XModuleFields'> | —
<class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | <class 'xblock.core.XBlock'> | —
<class 'xmodule.studio_editable.StudioEditableDescriptor'> | <class 'xmodule.studio_editable.StudioEditableModule'> | <class 'xmodule.studio_editable.StudioEditableBlock'> | —
